### PR TITLE
Affiliates Feature

### PR DESCRIPTION
**Pull request by mbareta**

### CODE REVIEW

File: lms/djangoapps/affiliates/views.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from django.shortcuts import render
+from django.http import Http404
+from lms.envs.common import STATE_CHOICES
+from django_countries import countries
+from edxmako.shortcuts import render_to_response, render_to_string
+from .models import AffiliateEntity
+
+# Create your views here.
+def index(request):
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/views.py]: Why use functional instead of class-based views?

**Comment by mbareta** [comment]
TODO:
- permissions
- ccx permissions

File: lms/djangoapps/affiliates/models.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    yield course
+
+
+@receiver(post_save, sender=AffiliateMembership, dispatch_uid="add_affiliate_course_enrollments")
+def add_affiliate_course_enrollments(sender, instance, **kwargs):
+    'Allow staff or instructor access to affiliate member into all affiliate courses if they are staff or instructor member.'
+    if not instance.role == 'ccx_coach':
+        for ccx in instance.affiliate.courses:
+            ccx_locator = CCXLocator.from_course_locator(ccx.course_id, ccx.id)
+            with ccx_course(ccx_locator) as course:
```
Comment (mbareta) [inline lms/djangoapps/affiliates/models.py]: Refactor to not use context manager.

File: common/djangoapps/student/views.py [hunk 3]
```diff
@@ -731,8 +596,7 @@ def dashboard(request):
         # author
         if hasattr(course.id, 'ccx'):
             ccx = CustomCourseForEdX.objects.get(pk=course.id.ccx)
-            author = ccx.coach
-            course_enrollment.author = author.profile.name
+            course_enrollment.author = ccx.coach.profile.affiliate.name
```
Comment (mbareta) [inline common/djangoapps/student/views.py]: Consider making ccx.affiliate method for easy fetching of affiliates.

File: lms/djangoapps/affiliates/decorators.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    """Only allow access to global staff or affiliate staff user(Program Director)"""
+    @wraps(function)
+    def wrapped_view(request, *args, **kwargs):
+        """Wrapper for the view function."""
+        if request.user.is_anonymous():
+            return HttpResponseNotFound()
+        elif request.user.is_staff:
+            return function(request, *args, **kwargs)
+        else:
+            has_pd_role_in_affiliate = AffiliateMembership.objects.filter(member=request.user, affiliate_id=kwargs['slug'], role='staff').exists()
```
Comment (mbareta) [inline lms/djangoapps/affiliates/decorators.py]: DRY this up. We don't need 2 queries for each role. We can look if role is in array.

File: lms/djangoapps/affiliates/management/commands/migrate_affiliates.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            image_url = get_profile_image_urls_for_user(user)['full']
+            self.stdout.write(image_url)
+
+            image_file = None
+
+            try:
+                if 's3' in image_url:
+                    file_data = urllib2.urlopen(image_url).read()
+                    image_file = ContentFile(file_data)
+                else:
```
Comment (mbareta) [inline lms/djangoapps/affiliates/management/commands/migrate_affiliates.py]: We should fix this to try and locate the file on the disk for easier migration.

Comment (MiroDojkic) [reply lms/djangoapps/affiliates/management/commands/migrate_affiliates.py]: What's the status on this?

File: lms/djangoapps/affiliates/management/commands/migrate_affiliates.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                    zipcode=user.profile.zipcode,
+                    facebook=user.profile.facebook_link,
+                    twitter=user.profile.twitter_link,
+                    linkedin=user.profile.linkedin_link,
+                    state=user.profile.state,
+                    country=user.profile.country,
+                )
+
+                if image_file:
+                    affiliate.image.save(image_url.split('/')[-1].split('?')[0], image_file, save=True)
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/management/commands/migrate_affiliates.py]: Could you please separate image_url transformation out of the function call to make this a bit more readable?

File: lms/djangoapps/affiliates/models.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    country = CountryField(blank=True, null=True)
+
+    image = models.ImageField(upload_to=user_directory_path, null=True, blank=True)
+
+    members = models.ManyToManyField(User, through='AffiliateMembership')
+
+    def save(self, *args, **kwargs):
+        slug = slugify(self.name)
+
+        if AffiliateEntity.objects.filter(slug=slug).exclude(pk=self.pk).exists():
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/models.py]: This would look better if separated from if statement.

File: lms/djangoapps/affiliates/models.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    if not instance.role == 'ccx_coach':
+        for ccx in instance.affiliate.courses:
+            ccx_locator = CCXLocator.from_course_locator(ccx.course_id, ccx.id)
+            course = get_course_by_id(ccx_locator)
+
+            try:
+                with transaction.atomic():
+                    allow_access(course, instance.member, instance.role, False)
+            except IntegrityError:
+                print 'IntegrityError: Allow access failed.'
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/models.py:154]: Shouldn't we handle this error differently?

File: lms/djangoapps/affiliates/models.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    if instance.role == 'staff' or instance.role == 'instructor':
+        for course_overview in course_overviews:
+            course_id = course_overview.id
+            course = get_course_by_id(course_id)
+
+            try:
+                with transaction.atomic():
+                    allow_access(course, instance.member, 'ccx_coach', False)
+            except IntegrityError:
+                print 'IntegrityError: CCX coach failed.'
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/models.py:168]: Also this.

File: lms/djangoapps/affiliates/views.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+from django.http import Http404
+from lms.envs.common import STATE_CHOICES
+from django_countries import countries
+from edxmako.shortcuts import render_to_response, render_to_string
+from .models import AffiliateEntity, AffiliateMembership
+from django.contrib.auth.models import User
+from lms.djangoapps.instructor.views.tools import get_student_from_identifier
+from .decorators import only_program_director, only_staff
+
+# Create your views here.
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/views.py]: Could you remove generated comments?

File: lms/djangoapps/affiliates/views.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        'member_id': member_id
+    }
+
+    AffiliateMembership.objects.filter(**params).delete()
+
+    return redirect('affiliates:edit', slug=slug)
+
+
+def is_program_director(user, affiliate):
+    if user.is_anonymous():
```
Comment (MiroDojkic) [inline lms/djangoapps/affiliates/views.py:175]: I think user.is_staff will return false if the user is anonymous

File: lms/djangoapps/affiliates/management/commands/migrate_affiliates.py (hunk 5 above)
Comment (mbareta) [reply lms/djangoapps/affiliates/management/commands/migrate_affiliates.py]: Still's missing.

File: lms/djangoapps/affiliates/models.py (hunk 8 above)
Comment (mbareta) [reply lms/djangoapps/affiliates/models.py:154]: Djangooooo... Sometimes works, sometimes throws this silly error. I handle it silently so it doesn't roll back the whole transaction.

File: lms/djangoapps/affiliates/models.py (hunk 9 above)
Comment (mbareta) [reply lms/djangoapps/affiliates/models.py:168]: Also same issue.

File: lms/djangoapps/affiliates/views.py (hunk 10 above)
Comment (mbareta) [reply lms/djangoapps/affiliates/views.py]: Of course.

File: lms/djangoapps/affiliates/views.py (hunk 11 above)
Comment (mbareta) [reply lms/djangoapps/affiliates/views.py:175]: It will and then it will try to find AffiliateMembership and blow up.

**Pull request merged**